File: android/app/src/main/java/com/plantora/billing/ui/sales/SalesViewModel.kt
package com.plantora.billing.ui.sales

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.plantora.billing.data.AuthState
import com.plantora.billing.data.BillRepository
import com.plantora.billing.data.ExpenseRepository
import com.plantora.billing.data.SalespersonRepository
import com.plantora.billing.data.SessionRepository
import com.plantora.billing.data.local.AppPreferences
import com.plantora.billing.R
import com.plantora.billing.data.remote.friendlyError
import com.plantora.billing.i18n.UiText
import com.plantora.billing.domain.BillListEntry
import com.plantora.billing.domain.DaySummary
import com.plantora.billing.domain.ExpenseCategory
import com.plantora.billing.domain.Money
import com.plantora.billing.domain.Role
import com.plantora.billing.domain.Salesperson
import com.plantora.billing.domain.toApiDate
import com.plantora.billing.domain.todayInShopZone
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate
import javax.inject.Inject

private const val PAGE = 20

/** State for the add/edit expense sheet. id == null → create. */
data class ExpenseEditor(
    val id: String? = null,
    val amount: String = "",
    /** Selected expense category id. Required to save (replaces free-text reason). */
    val categoryId: String? = null,
    val note: String = "",
    /** "cash" (out of the drawer) or "upi". */
    val paymentMethod: String = "cash",
    val saving: Boolean = false,
    val error: String? = null,
) {
    val canSave: Boolean get() = Money.parse(amount).isPositive() && categoryId != null && !saving
}

/** A salesperson and their sales total for the selected day (leaderboard row). */
data class StaffSales(val salesperson: Salesperson, val sales: Money)

data class SalesUiState(
    val date: LocalDate = todayInShopZone(),
    val isOwner: Boolean = false,
    val staff: List<Salesperson> = emptyList(),
    val staffSales: List<StaffSales> = emptyList(),
    val selectedStaffId: String? = null,
    val summaryLoading: Boolean = true,
    val summary: DaySummary? = null,
    /** Device preference: show cash in hand as a running all-time total. */
    val cashInHandCumulative: Boolean = false,
    val error: String? = null,
    val bills: List<BillListEntry> = emptyList(),
    val billsLoading: Boolean = true,
    val loadingMore: Boolean = false,
    val hasMore: Boolean = false,
    val expenseEditor: ExpenseEditor? = null,
    /** The shop's expense categories, for the add/edit picker. */
    val expenseCategories: List<ExpenseCategory> = emptyList(),
    /** Bill-number search box (digits only). Non-blank → the list shows search results
     *  across all history instead of the selected day. */
    val searchNo: String = "",
    val message: UiText? = null,
) {
    val isToday: Boolean get() = date == todayInShopZone()
    val isSearching: Boolean get() = searchNo.isNotBlank()
    /** Selected salesperson's email, or null for "all staff" (localized in the UI). */
    val selectedStaffEmail: String?
        get() = staff.find { it.id == selectedStaffId }?.email
}

@HiltViewModel
class SalesViewModel @Inject constructor(
    private val billRepo: BillRepository,
    private val expenseRepo: ExpenseRepository,
    private val salespersonRepo: SalespersonRepository,
    session: SessionRepository,
    prefs: AppPreferences,
) : ViewModel() {

    private val _ui = MutableStateFlow(SalesUiState())
    val ui: StateFlow<SalesUiState> = _ui.asStateFlow()

    init {
        val owner = (session.state.value as? AuthState.Authenticated)?.user?.role == Role.MANAGER
        _ui.update { it.copy(isOwner = owner) }
        // Keep the running-cash-in-hand display preference live.
        viewModelScope.launch {
            prefs.cashInHandCumulative.collect { v -> _ui.update { it.copy(cashInHandCumulative = v) } }
        }
        if (owner) loadStaff()
        // NOTE: the first/refresh load() is driven by the screen's resume effect
        // (see SalesScreen). The bottom nav saves/restores this screen's state, so
        // relying on init alone would show stale data — a bill made on the Bill tab
        // wouldn't appear until the VM was recreated. Reloading on resume keeps the
        // sales list and summary current every time the tab is opened.
    }

    private fun loadStaff() {
        viewModelScope.launch {
            runCatching { salespersonRepo.list() }
                .onSuccess { list -> _ui.update { it.copy(staff = list) }; refreshLeaderboard() }
        }
    }

    /**
     * Rank salespeople by their sales for the selected day so the owner can see who
     * sold the most. Fans out one day-summary request per staff member (small N).
     */
    private fun refreshLeaderboard() {
        val state = _ui.value
        if (!state.isOwner || state.staff.isEmpty()) return
        val date = state.date.toApiDate()
        val staff = state.staff
        viewModelScope.launch {
            val rows = staff.map { sp ->
                val sales = runCatching { billRepo.summary(date, sp.id).totalSales }.getOrDefault(Money.ZERO)
                StaffSales(sp, sales)
            }.sortedByDescending { it.sales }
            _ui.update { it.copy(staffSales = rows.filter { r -> r.sales.isPositive() }) }
        }
    }

    fun load() {
        // A bill-number search owns the list while it's active; the tab's resume refresh
        // must not wipe it. The search results are (re)driven by setSearchNo instead.
        if (_ui.value.isSearching) { runSearch(); return }
        val date = _ui.value.date.toApiDate()
        val staff = _ui.value.selectedStaffId
        _ui.update { it.copy(summaryLoading = true, billsLoading = true, error = null) }
        viewModelScope.launch {
            runCatching { billRepo.summary(date, staff) }
                .onSuccess { s -> _ui.update { it.copy(summaryLoading = false, summary = s) } }
                .onFailure { e -> _ui.update { it.copy(summaryLoading = false, error = friendlyError(e)) } }
        }
        viewModelScope.launch {
            runCatching { billRepo.list(date = date, createdBy = staff, limit = PAGE, offset = 0) }
                .onSuccess { p -> _ui.update { it.copy(billsLoading = false, bills = p.items, hasMore = p.hasMore) } }
                .onFailure { e -> _ui.update { it.copy(billsLoading = false, error = friendlyError(e)) } }
        }
        refreshLeaderboard()
    }

    /** Bill-number search: type digits to find a bill across all history; blank clears. */
    fun setSearchNo(v: String) {
        val digits = v.filter { it.isDigit() }.take(9)
        _ui.update { it.copy(searchNo = digits) }
        if (digits.isBlank()) load() else runSearch()
    }

    private fun runSearch() {
        val query = _ui.value.searchNo
        val no = query.toIntOrNull()
        _ui.update { it.copy(billsLoading = true, hasMore = false, error = null) }
        viewModelScope.launch {
            runCatching { billRepo.list(date = null, createdBy = null, billNo = no, limit = PAGE, offset = 0) }
                // Ignore a stale response if the query moved on while we were fetching.
                .onSuccess { p -> if (_ui.value.searchNo == query) _ui.update { it.copy(billsLoading = false, bills = p.items, hasMore = false) } }
                .onFailure { e -> if (_ui.value.searchNo == query) _ui.update { it.copy(billsLoading = false, error = friendlyError(e)) } }
        }
    }

    fun loadMore() {
        val state = _ui.value
        if (state.isSearching || state.loadingMore || !state.hasMore) return
        _ui.update { it.copy(loadingMore = true) }
        viewModelScope.launch {
            runCatching { billRepo.list(date = state.date.toApiDate(), createdBy = state.selectedStaffId, limit = PAGE, offset = state.bills.size) }
                .onSuccess { p -> _ui.update { it.copy(loadingMore = false, bills = it.bills + p.items, hasMore = p.hasMore) } }
                .onFailure { e -> _ui.update { it.copy(loadingMore = false, message = UiText.err(e, R.string.err_generic)) } }
        }
    }

    fun changeDate(date: LocalDate) { _ui.update { it.copy(date = date) }; load() }
    fun goToPreviousDay() = changeDate(_ui.value.date.minusDays(1))
    fun goToNextDay() {
        val next = _ui.value.date.plusDays(1)
        if (!next.isAfter(todayInShopZone())) changeDate(next)
    }

    fun selectStaff(id: String?) { _ui.update { it.copy(selectedStaffId = id) }; load() }

    // ── Expense editor ──
    fun openCreateExpense() {
        _ui.update { it.copy(expenseEditor = ExpenseEditor()) }
        loadExpenseCategories()
    }
    fun openEditExpense(id: String, amount: Money, categoryId: String?, note: String?, paymentMethod: String) {
        _ui.update {
            it.copy(expenseEditor = ExpenseEditor(id = id, amount = amount.toWire(), categoryId = categoryId, note = note.orEmpty(), paymentMethod = paymentMethod))
        }
        loadExpenseCategories()
    }
    fun closeExpenseEditor() = _ui.update { it.copy(expenseEditor = null) }
    fun setExpenseAmount(v: String) = _ui.update { it.copy(expenseEditor = it.expenseEditor?.copy(amount = v, error = null)) }
    fun setExpenseCategory(id: String?) = _ui.update { it.copy(expenseEditor = it.expenseEditor?.copy(categoryId = id, error = null)) }
    fun setExpenseNote(v: String) = _ui.update { it.copy(expenseEditor = it.expenseEditor?.copy(note = v, error = null)) }
    fun setExpenseMethod(v: String) = _ui.update { it.copy(expenseEditor = it.expenseEditor?.copy(paymentMethod = v, error = null)) }

    private fun loadExpenseCategories() {
        viewModelScope.launch {
            runCatching { expenseRepo.listCategories() }
                .onSuccess { cats -> _ui.update { it.copy(expenseCategories = cats) } }
        }
    }

    /** Manager-only: create a category and select it in the open editor. */
    fun addExpenseCategory(name: String) {
        if (name.isBlank()) return
        viewModelScope.launch {
            runCatching { expenseRepo.createCategory(name) }
                .onSuccess { created ->
                    _ui.update {
                        it.copy(
                            expenseCategories = (it.expenseCategories + created).sortedBy { c -> c.name.lowercase() },
                            expenseEditor = it.expenseEditor?.copy(categoryId = created.id, error = null),
                        )
                    }
                }
                .onFailure { e -> _ui.update { it.copy(expenseEditor = it.expenseEditor?.copy(error = friendlyError(e))) } }
        }
    }

    fun saveExpense() {
        val editor = _ui.value.expenseEditor ?: return
        if (!editor.canSave) return
        val categoryId = editor.categoryId ?: return
        _ui.update { it.copy(expenseEditor = editor.copy(saving = true, error = null)) }
        viewModelScope.launch {
            val amount = Money.parse(editor.amount)
            val note = editor.note.trim().ifBlank { null }
            val result = runCatching {
                if (editor.id != null) expenseRepo.update(editor.id, amount, categoryId, note, editor.paymentMethod)
                else expenseRepo.add(amount, categoryId, note, editor.paymentMethod)
            }
            result
                .onSuccess { _ui.update { it.copy(expenseEditor = null) }; load() }
                .onFailure { e -> _ui.update { it.copy(expenseEditor = editor.copy(saving = false, error = friendlyError(e))) } }
        }
    }

    fun deleteExpense(id: String) {
        viewModelScope.launch {
            runCatching { expenseRepo.delete(id) }.onSuccess { load() }
                .onFailure { e -> _ui.update { it.copy(message = UiText.err(e, R.string.err_generic)) } }
        }
    }

    fun deleteBill(id: String) {
        viewModelScope.launch {
            runCatching { billRepo.delete(id) }
                .onSuccess { _ui.update { it.copy(message = UiText.res(R.string.vm_bill_deleted)) }; load() }
                .onFailure { e -> _ui.update { it.copy(message = UiText.err(e, R.string.err_generic)) } }
        }
    }

    fun dismissMessage() = _ui.update { it.copy(message = null) }
}

File: Plantbill/Plantbill/Features/Sales/SalesViewModel.swift
import Foundation
import Observation

private let pageSize = 20

/// A salesperson and their sales total for the selected day (leaderboard row).
struct StaffSales: Identifiable, Equatable {
    let salesperson: Salesperson
    let sales: Money
    var id: UUID { salesperson.id }
}

/// State for the add/edit expense sheet. `id == nil` → create.
///
/// The backend's expense record only stores ONE payment method (`cash` or
/// `upi` — there's no split field on it, unlike a bill). "Split" is
/// therefore only offered when creating a new expense: on save it becomes
/// TWO separate expense rows (a cash one and a UPI one) whose amounts add up
/// to what was entered — each still valid against the existing schema, and
/// the day summary's cash/UPI expense totals (which sum per-row) come out
/// correct either way. Editing an existing row is always single-method,
/// since it's always one of those two rows.
struct ExpenseEditor: Equatable {
    var id: UUID? = nil
    var amount: String = ""
    var reason: String = ""
    /// "cash", "upi", or (create-only) "split".
    var paymentMethod: String = "cash"
    /// The cash portion when `paymentMethod == "split"`; the UPI portion is
    /// the remainder of `amount`.
    var splitCashText: String = ""
    var saving = false
    var error: String? = nil

    var amountMoney: Money { Money.parse(amount) }
    var splitCashMoney: Money { Money.parse(splitCashText) }
    var splitUpiMoney: Money {
        let remainder = amountMoney - splitCashMoney
        return remainder.isNegative ? .zero : remainder
    }

    var canSave: Bool {
        guard amountMoney.isPositive, !reason.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty, !saving else {
            return false
        }
        if paymentMethod == "split" {
            return splitCashMoney.isPositive && splitCashMoney <= amountMoney
        }
        return true
    }
}

@Observable
@MainActor
final class SalesViewModel {
    enum SummaryState: Equatable {
        case loading
        case loaded(DaySummary)
        case error(String)

        static func == (lhs: SummaryState, rhs: SummaryState) -> Bool {
            switch (lhs, rhs) {
            case (.loading, .loading): return true
            case (.loaded(let a), .loaded(let b)): return a.date == b.date && a.totalSales == b.totalSales
            case (.error(let a), .error(let b)): return a == b
            default: return false
            }
        }
    }

    let isManager: Bool

    private(set) var selectedDate: Date = ShopCalendar.today()
    private(set) var staff: [Salesperson] = []
    private(set) var staffSales: [StaffSales] = []
    private(set) var selectedStaffId: UUID?

    private(set) var summaryState: SummaryState = .loading
    private(set) var bills: [BillListEntry] = []
    private(set) var billsLoading = true
    private(set) var loadingMore = false
    private(set) var hasMore = false

    var expenseEditor: ExpenseEditor?
    var message: String?

    init(isManager: Bool) {
        self.isManager = isManager
    }

    var isToday: Bool { ShopCalendar.isToday(selectedDate) }
    var selectedStaffEmail: String? { staff.first { $0.id == selectedStaffId }?.email }

    func onAppear() async {
        if isManager && staff.isEmpty {
            await loadStaff()
        }
        await load()
    }

    private func loadStaff() async {
        do {
            staff = try await APIClient.shared.send(Endpoint(path: "shop/users"))
        } catch {
            staff = []
        }
    }

    /// Ranks salespeople by their sales for the selected day (fans out one
    /// day-summary request per staff member — small N).
    private func refreshLeaderboard() async {
        guard isManager, !staff.isEmpty else { return }
        let dateString = ShopCalendar.apiDateString(selectedDate)
        var rows: [StaffSales] = []
        for sp in staff {
            let query = [URLQueryItem(name: "date", value: dateString), URLQueryItem(name: "created_by", value: sp.id.uuidString)]
            if let summary: DaySummary = try? await APIClient.shared.send(Endpoint(path: "bills/summary/today", queryItems: query)) {
                rows.append(StaffSales(salesperson: sp, sales: summary.totalSalesMoney))
            }
        }
        staffSales = rows.filter { $0.sales.isPositive }.sorted { $0.sales > $1.sales }
    }

    func load() async {
        summaryState = .loading
        billsLoading = true

        let dateString = ShopCalendar.apiDateString(selectedDate)
        var mutableSummaryQuery = [URLQueryItem(name: "date", value: dateString)]
        if let selectedStaffId { mutableSummaryQuery.append(URLQueryItem(name: "created_by", value: selectedStaffId.uuidString)) }
        let summaryQuery = mutableSummaryQuery

        async let summaryTask: Void = loadSummary(query: summaryQuery)
        async let billsTask: Void = loadBills(dateString: dateString, reset: true)
        async let leaderboardTask: Void = refreshLeaderboard()
        _ = await (summaryTask, billsTask, leaderboardTask)
    }

    private func loadSummary(query: [URLQueryItem]) async {
        do {
            let summary: DaySummary = try await APIClient.shared.send(Endpoint(path: "bills/summary/today", queryItems: query))
            summaryState = .loaded(summary)
        } catch let error as APIError {
            summaryState = .error(error.userMessage)
        } catch {
            summaryState = .error(APIError.unknown.userMessage)
        }
    }

    private func loadBills(dateString: String, reset: Bool) async {
        do {
            var query = [
                URLQueryItem(name: "date_from", value: dateString),
                URLQueryItem(name: "date_to", value: dateString),
                URLQueryItem(name: "limit", value: "\(pageSize)"),
                URLQueryItem(name: "offset", value: "\(reset ? 0 : bills.count)"),
            ]
            if let selectedStaffId { query.append(URLQueryItem(name: "created_by", value: selectedStaffId.uuidString)) }
            let page: BillListPage = try await APIClient.shared.send(Endpoint(path: "bills", queryItems: query))
            bills = reset ? page.items : bills + page.items
            hasMore = page.hasMore
            billsLoading = false
            loadingMore = false
        } catch let error as APIError {
            billsLoading = false
            loadingMore = false
            message = error.userMessage
        } catch {
            billsLoading = false
            loadingMore = false
        }
    }

    func loadMore() async {
        guard !loadingMore, hasMore else { return }
        loadingMore = true
        await loadBills(dateString: ShopCalendar.apiDateString(selectedDate), reset: false)
    }

    func changeDate(_ date: Date) {
        selectedDate = ShopCalendar.calendar.startOfDay(for: date)
        Task { await load() }
    }

    func goToPreviousDay() {
        changeDate(ShopCalendar.calendar.date(byAdding: .day, value: -1, to: selectedDate) ?? selectedDate)
    }

    func goToNextDay() {
        let next = ShopCalendar.calendar.date(byAdding: .day, value: 1, to: selectedDate) ?? selectedDate
        if next <= ShopCalendar.today() { changeDate(next) }
    }

    func selectStaff(_ id: UUID?) {
        selectedStaffId = id
        Task { await load() }
    }

    // MARK: Expense editor

    func openCreateExpense() { expenseEditor = ExpenseEditor() }
    func openEditExpense(_ expense: Expense) {
        expenseEditor = ExpenseEditor(id: expense.id, amount: expense.amount, reason: expense.reason, paymentMethod: expense.paymentMethod)
    }
    func closeExpenseEditor() { expenseEditor = nil }

    func saveExpense() async {
        guard var editor = expenseEditor, editor.canSave else { return }
        editor.saving = true
        expenseEditor = editor
        let reason = editor.reason.trimmingCharacters(in: .whitespacesAndNewlines)
        do {
            if editor.id == nil, editor.paymentMethod == "split" {
                // No split field on the backend's expense model — post the
                // cash and UPI portions as two separate rows instead.
                if editor.splitCashMoney.isPositive {
                    let body = try APIClient.shared.encode(ExpenseRequest(amount: editor.splitCashMoney.toWire(), reason: reason, paymentMethod: "cash"))
                    let _: Expense = try await APIClient.shared.send(Endpoint(path: "expenses", method: .post, body: body))
                }
                if editor.splitUpiMoney.isPositive {
                    let body = try APIClient.shared.encode(ExpenseRequest(amount: editor.splitUpiMoney.toWire(), reason: reason, paymentMethod: "upi"))
                    let _: Expense = try await APIClient.shared.send(Endpoint(path: "expenses", method: .post, body: body))
                }
            } else {
                let body = try APIClient.shared.encode(ExpenseRequest(amount: editor.amountMoney.toWire(), reason: reason, paymentMethod: editor.paymentMethod))
                if let id = editor.id {
                    let _: Expense = try await APIClient.shared.send(Endpoint(path: "expenses/\(id)", method: .patch, body: body))
                } else {
                    let _: Expense = try await APIClient.shared.send(Endpoint(path: "expenses", method: .post, body: body))
                }
            }
            expenseEditor = nil
            await load()
        } catch let error as APIError {
            editor.saving = false
            editor.error = error.userMessage
            expenseEditor = editor
        } catch {
            editor.saving = false
            editor.error = APIError.unknown.userMessage
            expenseEditor = editor
        }
    }

    func deleteExpense(_ id: UUID) async {
        do {
            try await APIClient.shared.sendNoContent(Endpoint(path: "expenses/\(id)", method: .delete))
            await load()
        } catch let error as APIError {
            message = error.userMessage
        } catch {
            message = APIError.unknown.userMessage
        }
    }

    func dismissMessage() { message = nil }
}
